packages/transactions/src/useDraftDocument.ts

```diff
@@ -0,0 +1,121 @@
+ * while creating a new invoice.
+ */
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+export const createDraftStore = <T>(): UseStore<DraftStore<T>> => {
+  return zustand<DraftStore<T>>(set => ({
+    draft: null,
+    setDraft: draft => set(state => ({ ...state, draft })),
+  }));
+};
+
+interface DraftDocumentState<DocumentType> {
+  draft: DocumentType | null;
+  setDraft: (updatedDocument: DocumentType) => void;
+  save: () => void;
+}
+
+/**
+ * Hook to manage the client and server state relationship of a document we are editing.
+ * The "draft" is currently auto updated from whatever the server sends. The draft is then
+ * edited and saved 'at some point' in the future. When saving, the query is invalidated
+ * and the data is refetched from the server so that the server cache is refreshed and
+ * kept in sync.
+ *
+ * Improvements/TODO
+ *
+ * - Merge/conflict resolution functionality for when data is received from the server. You
+ * could set some state that was "There are differences between your version and the latest
+ * server version, please revise".
+ *
+ * - Auto-save: Could do a timer, or a debounced/throttled mutation on the `setDraft` function.
+ *
+ * - Rollbacks and warnings when saving doesn't succeed.
+ *
+ * - Optimistically update the server side data so mutation returns/re-querying doesn't cause
+ * a re-render if they are still equal.
+ *
+ * - More parameters/return values - be able to pass through configs for useQuery etc and return the
+ * loading/error states.
+ *
+ * - Easier editing mechanism. I don't want to have to use `setDraft({...draft, [myField]: newValue})
+ * every time - especially if there is more complexity - i.e. a customer requisition editing days of
+ * stock which might effect many columns. Possibly the draft could be a more complex class instance or
+ * we use a more functional approach and have functions i.e. updateDaysOfStock(draft, newValue)
+ *
+ * - Might be nice/easy/better/worse? to be able to more globally access the current draft. For example,
+ * the draft hook will be used at the 'root' of some page, and will be required to be drilled down through
+ * components which want to know the current draft values. It's not too difficult to make the value 'global'
+ * i.e. can use context, or a simple state management tool like Jotai or Recoil or..
+ * - Just use redux :thinking: :shrug: :lol:
+ */
+
+type MutateFn<DocumentType> = (
+  updatedDocument: DocumentType
+) => Promise<DocumentType>;
+
+type QueryFn<DocumentType> = () => Promise<DocumentType>;
+
+type OnSuccessCallback<DocumentType> = (
+  data: DocumentType,
+  variables: Partial<DocumentType>
+) => void;
+
+export const useDraftDocument = <DocumentType>(
+  key: unknown[],
+  queryFn: QueryFn<DocumentType>,
+  mutateFn: MutateFn<DocumentType>,
+  onSuccess: OnSuccessCallback<DocumentType>,
+  useDraftState: UseStore<DraftStore<DocumentType>>,
+  placeholderData?: DocumentType
```

Comment:
Here i've added the props `onSuccess`, `useDraftState` and `placeholderData`.

- `onSuccess` is for helping with the changing of the url once a `new` record has been saved. Also invalidates all related queries in the react-query cache, generally. Couldn't think of a nice way to abstract this so just passing it in for now.
- `useDraftState` - passing in a zustand store to use as state for the draft. Again couldn't think of a nice way to abstract this such that only one store per entity is created. Maybe we could pass in some sort of `entity` string identifier, but not sure that's so greater
- placeholderData: As above, used to seed the `data` returned from `useQuery` for a `new` record